### PR TITLE
Add cluster guid in errorMessage field when provision or update cluster failed

### PR DESCRIPTION
**Pull request by t-rufang**
* Add cluster guid in error log area when provision or update cluster failed
* Make error log area hideable in provision and update cluster dialog
* Make error log area as console view

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java [hunk 1]
```diff
@@ -76,13 +76,13 @@
...
     protected JLabel workerMemoryLabel;
     protected JLabel workerNumberOfContainersLabel;
-    protected JTextField errorMessageField;
     protected JPanel provisionDialogPanel;
     protected JButton refreshButton;
     protected JLabel storageRootPathLabel;
     protected JComboBox sparkVersionComboBox;
     protected JLabel sparkVersionLabel;
     private JXHyperLinkWithUri jobQueueHyperLink;
+    protected JTextArea errorMessageTextArea;
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java]: Is it scrollable?

Comment (t-rufang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java]: Made it scrollable in the new commit.

---
In reply to: [210529202](https://github.com/Microsoft/azure-tools-for-java/pull/1879#discussion_r210529202) [](ancestors = 210529202)

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessClusterUpdateDialog.java [hunk 2]
```diff
@@ -55,17 +56,13 @@ public SparkServerlessClusterUpdateDialog(@NotNull SparkServerlessClusterNode cl
...
                 this, new IdeaSchedulers((Project)clusterNode.getProject()), cluster);
         this.getWindow().addWindowListener(new WindowAdapter() {
             @Override
             public void windowOpened(WindowEvent e) {
                 ctrlProvider.initialize()
-                        .subscribe(complete -> {}, err -> {
-                            log().warn(String.format("Can't get the cluster %s details: %s", cluster.getName(), err));
-                            errorMessageField.setText("Error Loading cluster details");
-                        });
+                        .subscribe(complete -> {}, err -> {});
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessClusterUpdateDialog.java]: `err` should be caught.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java [hunk 3]
```diff
@@ -230,7 +257,7 @@ protected void doOKAction() {
                     // TODO: replace load with refreshWithoutAsync
                     adlAccountNode.load(false);
                     super.doOKAction();
-                }, err -> errorMessageField.setText(String.format("Provision failed. %s", err.getMessage())));
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java]: log a warning here.

File: Utils/hdinsight-node-common/src/com/microsoft/azure/sparkserverless/serverexplore/SparkServerlessClusterProvisionCtrlProvider.java [hunk 4]
```diff
@@ -139,7 +129,18 @@ private SparkServerlessClusterProvisionSettingsModel provisionCluster(
...
                 .observeOn(ideSchedulers.processBarVisibleAsync("Provisioning cluster..."))
                 .map(toUpdate -> toUpdate.setErrorMessage(null))
-                .map(toUpdate -> provisionCluster(toUpdate))
+                .flatMap(toUpdate ->
+                        buildCluster(toUpdate)
+                                .doOnNext(cluster -> toUpdate.setClusterGuid(cluster.getGuid()))
+                                .flatMap(cluster -> cluster.provision())
+                                .map(cluster -> toUpdate)
+                                .onErrorReturn(err -> {
+                                    String errMsg = String.format("%s Error: %s.\nCluster guid: %s\n",
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/sparkserverless/serverexplore/SparkServerlessClusterProvisionCtrlProvider.java]: `Error:` -> `ERROR`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessClusterUpdateDialog.java [hunk 5]
```diff
@@ -112,6 +109,11 @@ public void setData(@NotNull SparkServerlessClusterProvisionSettingsModel data)
         workerMemoryField.setText(String.valueOf(data.getWorkerMemory()));
         workerNumberOfContainersField.setText(String.valueOf(data.getWorkerNumberOfContainers()));
 
-        errorMessageField.setText(data.getErrorMessage());
+        if (!StringUtils.isEmpty(data.getErrorMessage())) {
+            if (!errorMessageDecorator.isExpanded()) {
+                errorMessageDecorator.setOn(true);
+            }
+            consoleViewPanel.print(data.getErrorMessage(), ConsoleViewContentType.ERROR_OUTPUT);
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessClusterUpdateDialog.java]: Better to also support `INFO_OUTPUT`

File: Utils/hdinsight-node-common/src/com/microsoft/azure/sparkserverless/serverexplore/SparkServerlessClusterUpdateCtrlProvider.java [hunk 6]
```diff
@@ -63,7 +65,12 @@ public SparkServerlessClusterUpdateCtrlProvider(
                 .flatMap(toUpdate ->
                         cluster.update(toUpdate.getWorkerNumberOfContainers())
                                 .map(cluster -> toUpdate)
-                                .onErrorReturn(err -> toUpdate.setErrorMessage(err.getMessage())))
+                                .onErrorReturn(err -> {
+                                    String errMsg = String.format("%s Error: %s.\nCluster guid: %s\n",
+                                            LocalDateTime.now(), err.getMessage(), cluster.getGuid());
+                                    log().error(errMsg + "\n" + err.toString());
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/sparkserverless/serverexplore/SparkServerlessClusterUpdateCtrlProvider.java]: `log().warning` or `log().info`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java [hunk 7]
```diff
@@ -192,10 +216,21 @@ private void updateCalculatedAU() {
                 Integer.valueOf(workerNumberOfContainersField.getText()))));
     }
 
+    protected void printErrorLog(@NotNull String errorLog, @Nullable String clusterGuid) {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java]: I suppose the function would be like:

```java
protected void printLogLine(@NotNull ConsoleViewContentType logLevel, @NotNull String log) {
    consoleViewPanel.print(LocalDateTime.now().toString() + " " + logLevel.toString().toUpperCase() + " " + log + "\n", logLevel);
}
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java [hunk 8]
```diff
@@ -192,10 +216,21 @@ private void updateCalculatedAU() {
...
+        if (!StringUtils.isEmpty(errorLog)) {
+            if (!errorMessageDecorator.isExpanded()) {
+                errorMessageDecorator.setOn(true);
+            }
+            consoleViewPanel.print(LocalDateTime.now().toString() + " ", ConsoleViewContentType.NORMAL_OUTPUT);
+            consoleViewPanel.print("ERROR: " + errorLog + "\n", ConsoleViewContentType.ERROR_OUTPUT);
+            consoleViewPanel.print("Cluster guid: " + clusterGuid + "\n", ConsoleViewContentType.NORMAL_OUTPUT);
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java]: Not mix the cluster guid with error logs.

File: Utils/hdinsight-node-common/src/com/microsoft/azure/sparkserverless/serverexplore/SparkServerlessClusterProvisionCtrlProvider.java [hunk 9]
```diff
@@ -139,7 +129,18 @@ private SparkServerlessClusterProvisionSettingsModel provisionCluster(
...
+                                            LocalDateTime.now(), err.getMessage(), toUpdate.getClusterGuid());
+                                    log().error(errMsg + "\n" + err.toString());
+                                    return toUpdate.setErrorMessage(errMsg);
+                                })
+                )
                 .observeOn(ideSchedulers.dispatchUIThread())
                 .doOnNext(controllableView::setData)
                 .filter(data -> StringUtils.isEmpty(data.getErrorMessage()));
```
Comment (djlan) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/sparkserverless/serverexplore/SparkServerlessClusterProvisionCtrlProvider.java]: NIT: Tab

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessClusterUpdateDialog.java [hunk 10]
```diff
@@ -112,6 +113,13 @@ public void setData(@NotNull SparkServerlessClusterProvisionSettingsModel data)
...
         workerNumberOfContainersField.setText(String.valueOf(data.getWorkerNumberOfContainers()));
 
-        errorMessageField.setText(data.getErrorMessage());
+        if (!StringUtils.isEmpty(data.getErrorMessage())) {
+            if (!errorMessageDecorator.isExpanded()) {
+                errorMessageDecorator.setOn(true);
+            }
+
+            printLogLine(ConsoleViewContentType.ERROR_OUTPUT, data.getErrorMessage());
+            printLogLine(ConsoleViewContentType.NORMAL_OUTPUT, "Cluster guid: " + cluster.getGuid());
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessClusterUpdateDialog.java]: Should the `NORMAL_OUTPUT` line be took out from only including error message condition?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java [hunk 11]
```diff
@@ -160,6 +170,20 @@ public void windowOpened(WindowEvent e) {
         });
     }
 
+    protected void setClusterNameSets() {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java]: The function name looks like a setter, but it's not.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java [hunk 12]
```diff
@@ -192,10 +216,21 @@ private void updateCalculatedAU() {
...
     @Override
     public void setData(@NotNull SparkServerlessClusterProvisionSettingsModel data) {
-        errorMessageField.setText(data.getErrorMessage());
+        if (!StringUtils.isEmpty(data.getErrorMessage())) {
+            if (!errorMessageDecorator.isExpanded()) {
+                errorMessageDecorator.setOn(true);
+            }
+
+            printLogLine(ConsoleViewContentType.ERROR_OUTPUT, data.getErrorMessage());
+            printLogLine(ConsoleViewContentType.NORMAL_OUTPUT, "Cluster guid: " + data.getClusterGuid());
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessProvisionDialog.java]: The same

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/common/azure/serverless/AzureSparkServerlessCluster.java [hunk 13]
```diff
@@ -786,13 +786,18 @@ public String getTenantId() {
...
     public Observable<Boolean> prepareStorageFolder(@NotNull String path) {
         return Observable.fromCallable(() -> {
-            String accessToken = getHttp().getAccessToken();
-            ADLStoreClient storeClient = ADLStoreClient.createClient(
-                    URI.create(this.getStorageAccount().getDefaultContainerOrRootPath()).getHost(), accessToken);
-            if (!storeClient.checkExists(path)) {
-                return storeClient.createDirectory(path);
+            try {
+                String accessToken = getHttp().getAccessToken();
+                ADLStoreClient storeClient = ADLStoreClient.createClient(
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/common/azure/serverless/AzureSparkServerlessCluster.java]: Use `try (storeClient = ...) {}` to close storeClient automatically.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessClusterUpdateDialog.java (hunk 10 above)
Comment (t-rufang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/sparkserverless/serverexplore/ui/SparkServerlessClusterUpdateDialog.java]: Taking the cluster guid log out of error message condition means cluster guid will also be shown when provision success. However, the cluster guid is only needed when provision failed to track the issue for backend team.

---
In reply to: [213200665](https://github.com/Microsoft/azure-tools-for-java/pull/1879#discussion_r213200665) [](ancestors = 213200665)

**Pull request merged**